File: src/main/lua/eyefuck.lua
--[[
Eyefuck.lua as i was saying in the eyefuck.rb i will **NOT** compile this
, go to eyefuck.rb for more info 
]]

-- ANSI colors
local RESET = "\27[0m"
local RED = "\27[31m"
local GREEN = "\27[32m"
local YELLOW = "\27[33m"
local BLUE = "\27[34m"
local CYAN = "\27[36m"
local WHITE = "\27[97m"

local EYF_V = 1.2
local TAPE_SIZE = 300000

local function main()
    if #arg < 1 then
        print(RED .. "Usage:" .. RESET .. " eyefuck <command> [file.eyf]")
        return
    end

    local mode = arg[1]

    if mode == "run" then
        if #arg < 2 then
            print(RED .. "Please specify a file to run." .. RESET)
            return
        end
        local file = arg[2]
        local f = io.open(file, "r")
        if not f then
            print("Error reading file")
            return
        end
        local code = f:read("*a")
        f:close()
        run_interpreter(code)
    elseif mode == "-i" or mode == "--i" or mode == "i" then
        start_repl()
    elseif mode == "help" or mode == "-help" or mode == "-h" or mode == "--h" or mode == "--help" then
        print(CYAN .. "Eyefuck HELP:" .. RESET)
        print(YELLOW .. "  eyefuck run <file.eyf>" .. RESET .. "  -> " .. GREEN .. "execute the Eyefuck file" .. RESET)
        print(YELLOW .. "  eyefuck -i" .. RESET .. "             -> " .. GREEN .. "interactive REPL mode" .. RESET)
        print(YELLOW .. "  eyefuck about" .. RESET .. "          -> " .. GREEN .. "information about this interpreter" .. RESET)
    elseif mode == "about" then
        print(CYAN .. "Eyefuck DEV 2025" .. RESET)
        print(GREEN .. "MIT license" .. RESET .. " see LICENSE for more information")
        print("Please help me motive by giving the repo a star")
        print(BLUE .. "github:" .. RESET .. " github.com/bandikaaking")
        print("crafted with " .. RED .. "<3" .. RESET .. " by " .. YELLOW .. "@Bandikaaking" .. RESET)
    elseif mode == "version" or mode == "--v" or mode == "--version" or mode == "-v" or mode == "v" or mode == "-version" then
        print("Current eyefuck version: " .. EYF_V)
    elseif mode == "ov" or mode == "-ov" or mode == "--ov" then
        print("Other Eyefuck versions: ")
        print("0.10: Started / added 2 instructions")
        print("0.11-0.43: Fixed many bugs, and edded 5 more instructions")
        print("1.0: Added syntax highliting")
        print("1.1: Fixed bugs")
        print("added more eyefuck modes / rewrited README.md")
    else
        print(RED .. "Unknown mode:" .. RESET .. " " .. mode)
    end
end

-- ---------------------------
-- Interactive REPL
-- ---------------------------
local function start_repl()
    print(CYAN .. "Eyefuck DEV 2025 - REPL" .. RESET)
    print("Type commands below, empty line to execute, Ctrl+C to exit")
    
    local code_lines = {}
    
    while true do
        io.write("$ ")
        io.flush()
        local line = io.read()
        
        if not line or line == "" then
            run_interpreter(table.concat(code_lines, "\n"))
            code_lines = {}
        else
            table.insert(code_lines, line)
        end
    end
end

-- ---------------------------
-- Eyefuck Interpreter
-- ---------------------------
local function run_interpreter(code)
    local tape = {}
    for i = 1, TAPE_SIZE do
        tape[i] = 0
    end
    local ptr = 1  -- Lua uses 1-based indexing
    local lines = {}
    for line in code:gmatch("[^\r\n]+") do
        table.insert(lines, line)
    end
    local loop_stack = {}
    local i = 1

    while i <= #lines do
        local line = lines[i]:match("^%s*(.-)%s*$")  -- trim
        
        -- remove comments after #
        local comment_pos = line:find("#")
        if comment_pos then
            line = line:sub(1, comment_pos - 1):match("^%s*(.-)%s*$") or ""
        end
        
        if line == "" then
            i = i + 1
            goto continue
        end

        if line == "^" then
            -- increment cell
            tape[ptr] = (tape[ptr] + 1) % 256
        elseif line == "v" then
            -- decrement cell
            tape[ptr] = (tape[ptr] - 1) % 256
        elseif line == ">" then
            -- move pointer right
            ptr = (ptr % TAPE_SIZE) + 1
        elseif line == "<" then
            -- move pointer left
            ptr = ptr == 1 and TAPE_SIZE or ptr - 1
        elseif line:sub(1, 3) == "bin" then
            -- set cell from binary
            local bin = line:sub(4):match("^%s*(.-)%s*$")
            local success, val = pcall(function()
                return tonumber(bin, 2)
            end)
            if success and val then
                tape[ptr] = math.floor(val) % 256
            else
                print("Invalid binary format")
                return
            end
        elseif line:sub(1, 3) == "col" then
            -- set text color from HEX
            local start_idx = line:find("%[")
            local end_idx = line:find("%]")
            if start_idx and end_idx and end_idx > start_idx + 1 then
                local hex = line:sub(start_idx + 1, end_idx - 1)
                local success, color_int = pcall(function()
                    return tonumber(hex, 16)
                end)
                if success and color_int then
                    local r = math.floor(color_int / 65536) % 256
                    local g = math.floor(color_int / 256) % 256
                    local b = color_int % 256
                    io.write(string.format("\27[38;2;%d;%d;%dm", r, g, b))
                else
                    print("Invalid HEX color")
                    return
                end
            end
        elseif line:sub(1, 5) == "load[" then
            -- load file
            local start_idx = line:find("%[")
            local end_idx = line:find("%]")
            if start_idx and end_idx and end_idx > start_idx + 1 then
                local filename = line:sub(start_idx + 1, end_idx - 1)
                local f = io.open(filename, "r")
                if f then
                    f:close()
                    tape[ptr] = 0
                else
                    print("Error loading file")
                    return
                end
            end
        elseif line == "," then
            -- read single byte input
            local input = io.read(1)
            if input then
                tape[ptr] = string.byte(input)
            end
        elseif line == "." then
            -- print cell as char
            io.write(string.char(tape[ptr]))
        elseif line == "loop[" then
            -- start loop
            table.insert(loop_stack, i)
        elseif line == "]" then
            -- end loop
            if tape[ptr] ~= 0 then
                if #loop_stack == 0 then
                    print("Unmatched ]")
                    return
                end
                i = loop_stack[#loop_stack]
            else
                if #loop_stack > 0 then
                    table.remove(loop_stack)
                end
            end
        else
            print(RED .. "error caught while parsing")
            print(RED .. "at line: " .. line)
            return
        end
        
        i = i + 1
        ::continue::
    end
    print()
end

-- Run main if this file is executed directly
if arg then
    main()
end
--good job Andrew i guess

File: src/main/rust/eyefuck.rs
/*

HEWHEHHHEHHEHEHEHEHH

i guess i'll rewrite it in RUST!!! anddddd this is `rustc` not `cargo`
i will write another Buidling for c# AND rust, but it is late and i am lazy

fact: you can actully

rustc eyefuck.rs -o eyefuck

generated an eyefuck.exe (or eyefuck.o)

and then paste it to c:\windows (or sudo cp eyefuck.o /usr/local/bin)

and BUMMM you now have eyefuck locally installed from the rust code

and i recomend doing the buildwin.bat, becuse it is more safer

and enjoy eyefuck
*/

use std::fs;
use std::io::{self, Write, BufRead, stdin};
use std::env;
use std::process;

// ANSI colors
const RESET: &str = "\x1b[0m";
const RED: &str = "\x1b[31m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const BLUE: &str = "\x1b[34m";
const CYAN: &str = "\x1b[36m";
const WHITE: &str = "\x1b[97m";

const EYF_V: f32 = 1.2;
const TAPE_SIZE: usize = 300000;

fn main() {
    let args: Vec<String> = env::args().collect();
    
    if args.len() < 2 {
        println!("{}Usage:{} eyefuck <command> [file.eyf]", RED, RESET);
        return;
    }

    let mode = &args[1];

    match mode.as_str() {
        "run" => {
            if args.len() < 3 {
                println!("{}Please specify a file to run.{}", RED, RESET);
                return;
            }
            let file = &args[2];
            let code = match fs::read_to_string(file) {
                Ok(content) => content,
                Err(e) => {
                    eprintln!("Error reading file: {}", e);
                    return;
                }
            };
            run_interpreter(&code);
        }
        "-i" | "--i" | "i" => {
            start_repl();
        }
        "help" | "-help" | "-h" | "--h" | "--help" => {
            println!("{}Eyefuck HELP:{}", CYAN, RESET);
            println!("{}  eyefuck run <file.eyf>{}  -> {}execute the Eyefuck file{}", YELLOW, RESET, GREEN, RESET);
            println!("{}  eyefuck -i{}             -> {}interactive REPL mode{}", YELLOW, RESET, GREEN, RESET);
            println!("{}  eyefuck about{}          -> {}information about this interpreter{}", YELLOW, RESET, GREEN, RESET);
        }
        "about" => {
            println!("{}Eyefuck DEV 2025{}", CYAN, RESET);
            println!("{}MIT license{} see LICENSE for more information", GREEN, RESET);
            println!("Please help me motive by giving the repo a star");
            println!("{}github:{} github.com/bandikaaking", BLUE, RESET);
            println!("crafted with {}<3{} by {}@Bandikaaking{}", RED, RESET, YELLOW, RESET);
        }
        "version" | "--v" | "--version" | "-v" | "v" | "-version" => {
            println!("Current eyefuck version: {}", EYF_V);
        }
        "ov" | "-ov" | "--ov" => {
            println!("Other Eyefuck versions: ");
            println!("0.10: Started / added 2 instructions");
            println!("0.11-0.43: Fixed many bugs, and edded 5 more instructions");
            println!("1.0: Added syntax highliting");
            println!("1.1: Fixed bugs");
            println!("added more eyefuck modes / rewrited README.md");
        }
        _ => {
            println!("{}Unknown mode:{} {}", RED, RESET, mode);
        }
    }
}

// ---------------------------
// Interactive REPL
// ---------------------------
fn start_repl() {
    println!("{}Eyefuck DEV 2025 - REPL{}", CYAN, RESET);
    println!("Type commands below, empty line to execute, Ctrl+C to exit");
    
    let stdin = io::stdin();
    let mut code_lines = Vec::new();
    
    loop {
        print!("$ ");
        io::stdout().flush().unwrap();
        
        let mut line = String::new();
        match stdin.lock().read_line(&mut line) {
            Ok(0) => break, // EOF
            Ok(_) => {
                let line = line.trim().to_string();
                if line.is_empty() {
                    run_interpreter(&code_lines.join("\n"));
                    code_lines.clear();
                    continue;
                }
                code_lines.push(line);
            }
            Err(e) => {
                eprintln!("Error reading input: {}", e);
                break;
            }
        }
    }
}

// ---------------------------
// Eyefuck Interpreter
// ---------------------------
fn run_interpreter(code: &str) {
    let mut tape = vec![0u8; TAPE_SIZE];
    let mut ptr = 0;
    let lines: Vec<&str> = code.lines().collect();
    let mut loop_stack: Vec<usize> = Vec::new();
    let mut i = 0;

    while i < lines.len() {
        let mut line = lines[i].trim();
        
        // remove comments after #
        if let Some(comment_pos) = line.find('#') {
            line = &line[..comment_pos].trim();
        }
        
        if line.is_empty() {
            i += 1;
            continue;
        }

        match line {
            "^" => {
                // increment cell
                tape[ptr] = tape[ptr].wrapping_add(1);
            }
            "v" => {
                // decrement cell
                tape[ptr] = tape[ptr].wrapping_sub(1);
            }
            ">" => {
                // move pointer right
                ptr = (ptr + 1) % TAPE_SIZE;
            }
            "<" => {
                // move pointer left
                ptr = if ptr == 0 { TAPE_SIZE - 1 } else { ptr - 1 };
            }
            _ if line.starts_with("bin") => {
                // set cell from binary
                let bin = line[3..].trim();
                match u8::from_str_radix(bin, 2) {
                    Ok(val) => tape[ptr] = val,
                    Err(e) => {
                        eprintln!("Invalid binary: {}", e);
                        process::exit(1);
                    }
                }
            }
            _ if line.starts_with("col") => {
                // set text color from HEX
                if let Some(start) = line.find('[') {
                    if let Some(end) = line.find(']') {
                        if end > start + 1 {
                            let hex = &line[start+1..end];
                            match u32::from_str_radix(hex, 16) {
                                Ok(color_int) => {
                                    let r = (color_int >> 16) & 0xFF;
                                    let g = (color_int >> 8) & 0xFF;
                                    let b = color_int & 0xFF;
                                    print!("\x1b[38;2;{};{};{}m", r, g, b);
                                }
                                Err(_) => {
                                    eprintln!("Invalid HEX color");
                                    process::exit(1);
                                }
                            }
                        }
                    }
                }
            }
            _ if line.starts_with("load[") => {
                // load file
                if let Some(start) = line.find('[') {
                    if let Some(end) = line.find(']') {
                        if end > start + 1 {
                            let filename = &line[start+1..end];
                            match fs::read(filename) {
                                Ok(_data) => {
                                    tape[ptr] = 0;
                                }
                                Err(e) => {
                                    eprintln!("Error loading file: {}", e);
                                    process::exit(1);
                                }
                            }
                        }
                    }
                }
            }
            "," => {
                // read single byte input
                let mut input = [0u8; 1];
                if let Ok(_) = io::stdin().read_exact(&mut input) {
                    tape[ptr] = input[0];
                }
            }
            "." => {
                // print cell as char
                print!("{}", tape[ptr] as char);
            }
            "loop[" => {
                // start loop
                loop_stack.push(i);
            }
            "]" => {
                // end loop
                if tape[ptr] != 0 {
                    if let Some(&loop_start) = loop_stack.last() {
                        i = loop_start;
                    } else {
                        eprintln!("Unmatched ]");
                        process::exit(1);
                    }
                } else {
                    loop_stack.pop();
                }
            }
            _ => {
                eprintln!("{}error caught while parsing", RED);
                eprintln!("{}at line: {}", RED, line);
                process::exit(1);
            }
        }
        
        i += 1;
    }
    println!();
}
//good job Andrew i guess


/* why am i doing this.... */